### DISCUSSION TITLE
Welcome to CnavasHack Discussions!

### DISCUSSION THREAD

**Discussion by castle-samj** [castle-samj/CnavasHack#6] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by wingedthing** [comment]
What methods are we going to use to decide if a domain is up? Here is what I have:

## For each polling cycle this would occur for each domain tracked:

1. Ping ip address. If this fails then we can be pretty sure the site is down, assuming we have the correct ip and the domain allows pings. 

2. If 1 is successful or the domain does not allow pings, then try to get around [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS), and send a http get request and listen for status codes.

3. If the domain doesn't allow cross-origin HTTP requests or the status returned was bad, then attempt to navigate to the domain using a headless browser.
    - If the domain is behind authentication, we can provide it or skip to step 4. 
    - When on the domain we should record the http status as well. 

4. If we can not navigate to the domain at all or the http status returned was bad. Check a downage site provided by the domain and either interface with their api (doubtful there will be one) or scrape the site for downage info.
    - If the downage site shows the domain as down, we fire off a msg to discord.
    - If the downage site says the domain is up, we add a timer flag to to the db entry for the domain that will track how many times in a row it has reached this point and fire a msg to discord after a certain amount of checks?
    - If there is no downage site or it is not reliable with its updates, we can fire a msg right away or like wise start tracking how many fails we had in a row and send msg after certain amount.

**Reply by castle-samj** [reply]
There should probably be some integrated latency to prevent false-negatives. I'm thinking:  
first attempt (t = 0) | retry at half-time (t = 30) | retry at half-time (t = 60)  
--- | --- | ---
| pingx1 = fail | pingx2 = fail | pingx3 = fail + init.error.msg |

**Reply by wingedthing** [reply]
That would work. Though upon further thinking, maybe ping isn't very useful other than in cases where we know it will be accepted? A lot domains will block them and even if we get a ping back that doesn't really tell us if the service we are concerned about is available. So no matter what, we are probably going to still have to ask the domain to service a request. Also, each request we make over the internet adds time to our result determination unless we decide that the ping response isn't integral to subsequent logic and we can then make it asynchronously, but that kind of defeats the purpose of using it as gate keeping conditional.

**Comment by wingedthing** [comment]
Check out [heroku](https://www.heroku.com/pricing) for hosting, they have a free tier that might cover everything we need.
